std-bits/table/src/main/java/org/enso/table/data/index/MultiValueKey.java

```diff
@@ -71,4 +74,120 @@ protected static Object FoldObject(Object value) {
+    // NULLs
+    if (thisValue == null) {
+      if (thatValue != null) {
+        return 1;
+      }
+      return 0;
+    }
+    if (thatValue == null) {
+      return -1;
+    }
+
+    // Booleans
+    if (thisValue instanceof Boolean && thatValue instanceof Boolean) {
+      boolean thisBool = (Boolean)thisValue;
+      boolean thatBool = (Boolean)thatValue;
+      if (thisBool == thatBool) {
+        return 0;
+      }
+      return thisBool ? 1 : -1;
+    }
+
+    // Long this
+    if (thisValue instanceof Long) {
+      Long thisLong = (Long)thisValue;
+      if (thatValue instanceof Long) {
+        return thisLong.compareTo((Long)thatValue);
+      }
+      if (thatValue instanceof Double) {
+        Double thatDouble = (Double)thatValue;
+        if (thisLong > thatDouble) {
+          return 1;
+        }
+        if (thisLong < thatDouble) {
+          return -1;
+        }
+        return 0;
+      }
+    }
+
+    // Double this
+    if (thisValue instanceof Double) {
+      Double thisDouble = (Double)thisValue;
+      if (thatValue instanceof Double) {
+        return thisDouble.compareTo((Double)thatValue);
+      }
+      if (thatValue instanceof Long) {
+        Long thatLong = (Long)thatValue;
+        if (thisDouble > thatLong) {
+          return 1;
+        }
+        if (thisDouble < thatLong) {
+          return -1;
+        }
+        return 0;
+      }
+    }
+
+    // Text
+    if (thisValue instanceof String && thatValue instanceof String) {
+      // Needs to use ICU comparisons
+      return ((String)thisValue).compareTo((String)thatValue);
+    }
+
+    // DateTimes
+    if (thisValue instanceof LocalDate) {
+      LocalDate thisDate = (LocalDate)thisValue;
+      if (thatValue instanceof LocalDate) {
+        return thisDate.compareTo((LocalDate)thatValue);
+      }
+      if (thatValue instanceof LocalDateTime) {
+        return thisDate.atStartOfDay().compareTo((LocalDateTime)thatValue);
+      }
+    }
+    if (thisValue instanceof LocalDateTime) {
+      LocalDateTime thisDateTime = (LocalDateTime)thisValue;
+      if (thatValue instanceof LocalDate) {
+        return thisDateTime.compareTo(((LocalDate)thatValue).atStartOfDay());
+      }
+      if (thatValue instanceof LocalDateTime) {
+        return thisDateTime.compareTo((LocalDateTime)thatValue);
+      }
+    }
+
+    // TimeOfDay
+    if (thisValue instanceof LocalTime) {
+      LocalTime thisTime = (LocalTime)thisValue;
+      if (thatValue instanceof LocalTime) {
+        return thisTime.compareTo((LocalTime)thatValue);
+      }
+    }
+
+    // Give Up!
+    throw new ClassCastException("Incomparable keys.");
+  }
+
+  @Override
+  public int compareTo(MultiValueKey that) {
+    if (that == null) {
+      throw new NullPointerException();
+    }
+
+    if (that.values.length != values.length) {
+      throw new ClassCastException("Incomparable keys.");
```

Comment:
Why is this a `ClassCastException`? I don't see any class casting. Shouldn't this be something like `IllegalArgumentException` or other?

Reply:
That's the exception thrown by the Comparable interface and so used here.

Reply:
Ah, I see. Weird, but if that's what the contract of `Comparable` specifies, then I'm not opposing.